ocaml/typing/mode.ml

```diff
@@ -1115,11 +1118,29 @@ module type Obj = sig
   val obj : const C.obj
 end
 
-let equate_from_submode submode m0 m1 =
-  match submode m0 m1 with
+let atomically op =
+  let log = ref S.empty_changes in
+  match op ~log with
+  | Ok _ as x ->
+    !append_changes log;
+    x
+  | Error _ as x ->
+    S.undo_changes !log;
+    x
+  [@@inline]
+
+let atomically' op =
```

Comment:
Maybe `append_local_log`? There's nothing particularly atomic here.

Reply:
Hmm I do hope to mean that `atomically'` is a special case of `atomically` that always succeeds. Do you have suggestions?

Reply:
But if it always succeeds, then there's nothing interestingly atomic about it. It's needed just because some functions work with a local log, instead of appending to the global one. That's why I suggested `append_local_log` or maybe `with_local_log`.

Reply:
I have renamed them to `with_log` and `try_with_log` - that should incorporate both of ours ideas.